pkg/agent/flexvolume/server.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2017 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+Some of the code below came from https://github.com/coreos/etcd-operator
+which also has the apache 2.0 license.
+*/
+
+// Package flexvolume to manage Kubernetes storage attach events.
+package flexvolume
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"net/rpc"
+	"os"
+	"path"
+
+	"k8s.io/client-go/rest"
+
+	"github.com/coreos/pkg/capnslog"
+	"github.com/rook/rook/pkg/clusterd"
+)
+
+const (
+	UnixSocketName           = ".rook.sock"
+	FlexvolumeVendor         = "rook.io"
+	FlexvolumeDriver         = "rook"
+	flexvolumeDriverFileName = "flexvolume"
+)
+
+var flexVolumeDriverDir = fmt.Sprintf("/flexmnt/%s~%s", FlexvolumeVendor, FlexvolumeDriver)
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "rook-flexvolume")
+
+// FlexvolumeServer start a unix domain socket server to interact with the flexvolume driver
+type FlexvolumeServer struct {
+	controller *FlexvolumeController
+	listener   net.Listener
+}
+
+// NewFlexvolumeServer creates an Flexvolume server
+func NewFlexvolumeServer(context *clusterd.Context, volumeAttachmentClient rest.Interface, manager VolumeManager) *FlexvolumeServer {
+	return &FlexvolumeServer{
+		controller: newFlexvolumeController(context, volumeAttachmentClient, manager),
+	}
+}
+
+// Start configures the flexvolume driver on the host and starts the unix domain socket server to communicate with the driver
+func (s *FlexvolumeServer) Start() error {
+	err := configureFlexVolume(flexvolumeDriverFileName, flexVolumeDriverDir)
+	if err != nil {
+		return fmt.Errorf("unable to configure flexvolume: %v", err)
+	}
+
+	err = rpc.Register(s.controller)
+	if err != nil {
+		return fmt.Errorf("unable to register rpc: %v", err)
+	}
+	logger.Info("Rook Flexvolume configured")
+
+	unixSocketFile := path.Join(flexVolumeDriverDir, path.Join(UnixSocketName)) // /flextmnt/rook.io~rook/.rook.sock
+
+	// remove unix socket if it existed previously
+	if _, err := os.Stat(unixSocketFile); !os.IsNotExist(err) {
+		logger.Info("Deleting unix domain socket file.")
+		os.Remove(unixSocketFile)
+	}
+
+	s.listener, err = net.Listen("unix", unixSocketFile)
+	if err != nil {
+		return fmt.Errorf("unable to listen at %s: %v", unixSocketFile, err)
+	}
+	go rpc.Accept(s.listener)
```

Comment:
should the `rpc.Accept()` call be moved to after the socket file permissions are set?

Reply:
sure